src/main/java/org/protege/owlapi/concurrent/WriteSafeOWLOntologyImpl.java

```diff
@@ -1318,4 +1297,861 @@ public boolean isTopEntity() {
         }
     } 
     
+    @Override
+    public void accept(OWLNamedObjectVisitor visitor) {
+        readLock.lock();
+        try {
+            delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <O> O accept(OWLNamedObjectVisitorEx<O> visitor) {
+        readLock.lock();
+        try {
+            return delegate.accept(visitor);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public void setOWLOntologyManager(OWLOntologyManager saveOntology) {
+        writeLock.lock();
+        try {
+            delegate.setOWLOntologyManager(saveOntology);
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    @Override
+    public void saveOntology() throws OWLOntologyStorageException {
+        writeLock.lock();
```

Comment:
Thanks for the clarifications Tim.  

I think the reader/writer locks make sense, but coming to this code with a fresh pair of eyes, I'm just nervous about changing things that could introduce deadlocks etc.  OWLOntology is fairly straightforward... it's other things like OWLOntologyManager that seem more complex and bothersome.

>Yes that is probably true.  This api was the dumping ground of
everything that did not have a place.  At one time I had hoped that the
reader-writer locks would find a home in the OWL api because I thought
that it was a simpler model than other solutions that existed at the time.

I think it would be better if this stuff is in the OWL API (I do of course understand the motivation for doing it and for it not being in the OWL API at the moment).  At the moment it's a maintenance headache (keeping up with new versions etc.).  I expect all of this will break with version 5.0.0 of the OWL API :(